pkg/apis/pubsub/v1alpha1/topic_types.go

```diff
@@ -0,0 +1,160 @@
+
+	// Topic is the ID of the Topic to create/use in Google Cloud Pub/Sub.
+	Topic string `json:"topic,omitempty"`
+
+	//PropagationPolicy defines how Topic controls the Cloud Pub/Sub topic for lifecycle changes.
+	PropagationPolicy PropagationPolicyType `json:"propagationPolicy,omitempty"`
+
+	// ServiceAccoutName is the name of the ServiceAccount that will be used to
+	// run the Topic Deployment.
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+}
+
+// PropagationPolicyType defines enum type for TopicPolicy
+type PropagationPolicyType string
+
+const (
+	// TopicPolicyCreateDelete defines the Cloud Pub/Sub topic management
+	// policy for creating topic if not present, and delete topic when the
+	// Topic resource is deleted.
+	TopicPolicyCreateDelete PropagationPolicyType = "CreateDelete"
+
+	// TopicPolicyCreateRestrictDelete defines the Cloud Pub/Sub topic
+	// management policy for creating topic if not present, and do not delete
+	// topic when the Topic resource is deleted.
+	TopicPolicyCreateRestrictDelete PropagationPolicyType = "CreateRestrictDelete"
+
+	// TopicPolicyRestrictCreateRestrictDelete defines the Cloud Pub/Sub topic
+	// management policy for only using existing topics, and do not delete
+	// topic when the Topic resource is deleted.
+	TopicPolicyRestrictCreateRestrictDelete PropagationPolicyType = "RestrictCreateRestrictDelete"
+)
+
+var topicCondSet = apis.NewLivingConditionSet(
+	TopicConditionAddressable,
+	TopicConditionTopicExists,
+	TopicConditionPublisherReady,
+)
+
+const (
+	// TopicConditionReady has status True when all subconditions below have
+	// been set to True.
+	TopicConditionReady = apis.ConditionReady
+
+	// TopicConditionAddressable has status true when this Topic meets the
+	// Addressable contract and has a non-empty hostname.
+	TopicConditionAddressable apis.ConditionType = "Addressable"
+
+	// TopicConditionTopicExists has status True when the Topic has had a
+	// Pub/Sub topic created for it.
+	TopicConditionTopicExists apis.ConditionType = "TopicExists"
+
+	// TopicConditionPublisherReady has status True when the Topic has had
+	// its publisher deployment created.
+	TopicConditionPublisherReady apis.ConditionType = "PublisherReady"
+)
+
+// TopicStatus represents the current state of a Topic.
+type TopicStatus struct {
+	// inherits duck/v1beta1 Status, which currently provides:
+	// * ObservedGeneration - the 'Generation' of the Service that was last processed by the controller.
+	// * Conditions - the latest available observations of a resource's current state.
+	duckv1beta1.Status `json:",inline"`
+
+	// Topic is Addressable. It currently exposes the endpoint as a
+	// fully-qualified DNS name which will distribute traffic over the
+	// provided targets from inside the cluster.
+	//
+	// It generally has the form {Topic}.{namespace}.svc.{cluster domain name}
+	duckv1alpha1.AddressStatus `json:",inline"`
+
+	// ProjectID is the resolved project ID in use by the Topic.
```

Comment:
Is this different from the projectID in the spec?

Reply:
project id in spec is optional.